conctr.agent.lib.nut

```diff
@@ -195,18 +232,296 @@ class Conctr {
+        }
+        _pubSubEndpoints = _formPubSubEndpointUrls(_app_id, _api_key, _device_id, _region, _env);
+        return _protocol
+    }
+
+
+    // 
     // Requests the device sends its current location to the agent or to Conctr
-    //
+    // 
     // @param  {Boolean} sendToConctr - If true the location will be sent to Conctr. If false, it will be cached on the agent.
-    //
+    // 
     function sendLocation(sendToConctr = true) {
 
         if (DEBUG) server.log("Conctr: requesting location be sent to " + (sendToConctr ? "conctr" : "agent"));
         _sender.send(LOCATION_REQ_EVENT, sendToConctr);
 
     }
-    
+
+
+    // 
+    // Publishes a message to a specific topic.
+    // 
+    // @param  {String/Array}   topics      List of Topics that message should be sent to
+    // @param  {[type]}         msg         Data to be sent to be published
+    // @param  {[String}        contentType Header specifying the content type of the msg
+    // @param  {Function}       cb          Function called on completion of publish request
+    // 
+    function publish(topics, msg, contentType = null, cb = null) {
+        local relativeUrl = ""
+        _publish(relativeUrl, { "topics": topics }, msg, contentType, cb);
+
+    }
+
+
+    // 
+    // Publishes a message to a specific device.
+    // @param  {String/Array}   deviceId     Device id(s) the message should be sent to
+    // @param  {[type]}         msg          Data to be sent to be published
+    // @param  {[type]}         contentType  Header specifying the content type of the msg
+    // @param  {Function}       cb           Function called on completion of publish request
+    // 
+    function publishToDevice(deviceIds, msg, contentType = null, cb = null) {
+        local relativeUrl = "/dev/";
+        _publish(relativeUrl, { "device_ids": deviceIds }, msg, contentType, cb);
+    }
+
+
+    // 
+    // Subscribe to a single/list of topics
+    // 
+    // @param  {Function}       cb     Function called on receipt of data
+    // @param  {Array/String}   topics String or Array of string topic names to subscribe to
+    // 
+    function subscribe(topics = [], cb = null) {
+
+        // Check if only callback is subscribed to
+        if (typeof topics == "function") {
+            cb = topics;
+            topics = [];
+        }
+
+        if (cb == null) {
+            throw "Conctr: callback to subscribe is a required parameter";
+        }
+
+        // If no topics to subscribe to found, subscribe to this devices id
+        if (topics.len() == 0) {
+            topics = ["dev/" + _device_id];
+        }
+
+        if (typeof topics != "array") {
+            topics = [topics];
+        }
+
+        local action = "subscribe";
+        local headers = {};
+        local payload = {};
+        local chunks = "";
+        local contentLength = null;
+        local reqTime = time();
+
+        // http done callback
+        local _doneCb = function(resp) {
```

Comment:
It would be a lot more readable if this function wasn't declared within the subscribe function but was a private method on the class.